sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java

```diff
@@ -193,16 +199,43 @@ public ShardCollectSource(Settings settings,
             );
         }
 
-
         // actual shards might be less if table is partitioned and a partition has been deleted meanwhile
-        int maxNumShards = normalizedPhase.routing().numShards(localNodeId);
-        ShardProjectorChain projectorChain = ShardProjectorChain.passThroughMerge(
-                normalizedPhase.jobId(),
-                maxNumShards,
-                normalizedPhase.projections(),
-                downstream,
-                projectorFactory,
-                jobCollectContext.queryPhaseRamAccountingContext());
+        final int maxNumShards = normalizedPhase.routing().numShards(localNodeId);
+        RowDownstream.Factory rowDownstreamFactory;
+        CompositeCollector.Builder builder = null;
+        boolean hasAnyShardProjections = false;
+        for (Projection projection : normalizedPhase.projections()) {
+            if (projection.requiredGranularity().ordinal() >= RowGranularity.SHARD.ordinal()) {
+                hasAnyShardProjections = true;
+                break;
+            }
+        }
+        if (hasAnyShardProjections) {
+            rowDownstreamFactory = new RowDownstream.Factory() {
```

Comment:
might make sense to have a non-anonymous implementation for this which is then also used in ShardProjectorChainTest